Visual C/C++ compiler support

cristianadam opened the pull request:
I picked only the compiler commits from: https://github.com/ccache/ccache/pull/162

The following commits I've adapted to the latest ccache C++ code:

375fe24: Add compiler_is_msvc() and MSVC specific option table.
7e01763: Add handling of /Fo option (replaces -o, but shall have no
space)
0c5cd25: Manage /E, /c equivalence. -g is gcc only. -O or /O is msvc
only.
4f61b59: MSVC send part of the error/warning messages to STDOUT, so
concat wit…
64449d6: use common code for gcc an cl response files.

I have tested this with CMake using the Ninja generator in Release mode. It seems to work good enough for having it in GitHub actions for [Qt Creator](https://github.com/qt-creator/qt-creator/actions).

- [x] Release build with Ninja
- [x] Debug / RelWithDebInfo build with Ninja (requires `/Z7` flag)
- [x] NMake Makefiles (requires usage of response files)
- [x] Precompile headers support (correct build without storage of pch files)
- [ ] Precompile headers support (detection and storage of pch files)
- [ ] Unittests

Fixes: https://github.com/ccache/ccache/issues/431

cristianadam commented on src/ccache.cpp:
```diff
@@ -2690,7 +2769,7 @@ cc_process_args(struct args* args,
     // Debugging is handled specially, so that we know if we can strip line
     // number info.
-    if (str_startswith(argv[i], "-g")) {
+    if (str_startswith(argv[i], "-g") && guessed_compiler != GUESSED_MSVC) {
```
MSVC has a `-/GZ` parameter, which would be taken into consideration here. But since the comparison is case sensitive I think it should be fine without it.

I'll do a test and update the PR.

jrosdahl commented:
> Debug build with Ninja (requires storage of pdb files)

[Knowledge level: 5 minutes of DuckDuckGoing.] Do I understand correctly that the above means that `/Zi` and `/ZI` should be made TOO_HARD until support has been implemented? Any other options that should be too hard?

> Precompile headers support (detection and storage of pch files)

Similarly: Should `/Yc` be too hard?

> Release build with NMake Makefiles (requires usage of response files)

Does this mean to write a parser similar to `args_init_from_gcc_atfile` but for a MSVC-specific format?

rvjr commented:
We are also using ccache for our Linux builds and we do need a solution for MSVC. I've been experimenting for some days now with the similar python based clcache and they also don't support /Yc and /Zi or /ZI. We haven't been using precompiled headers so far, but I switched a group of larger projects from /Zi to /Z7 and I didn't feel any disadvantage so far, so I think it is probably the easiest option to add /Zi, /ZI and /Yc into the TOO_HARD list for a first initial MSVC support.

jrosdahl commented:
@cristianadam: Are you still interested in working on this?

cristianadam commented:
> @cristianadam: Are you still interested in working on this?

Yes. I will be having some time to work on this next week.

jacquesg commented:
Support ```/Z7``` would be super useful.

schiele commented:
As suggested by @jrosdahl I created a pull request https://github.com/cristianadam/ccache/pull/1 to suggest updating that branch instead of the pull request running in parallel to this one. The pull request essentially forward-ports the changes from this branch to the current ccache master branch. In addition to that some trivial reformatting is taking place to make the clang-format test happy.

asheplyakov commented:
> As mentioned in my response to #162, a sibling "ccache for windows" project, run by others, would be great! Would you like to help with such an effort?

In fact there is such a project: https://github.com/mbitsnbites/buildcache
(Although it's not for windows only,  GCC and Linux are supported too)

jrosdahl commented:
> > As mentioned in my response to #162, a sibling "ccache for windows" project, run by others, would be great! Would you like to help with such an effort?
> 
> In fact there is such a project: https://github.com/mbitsnbites/buildcache
> (Although it's not for windows only, GCC and Linux are supported too)

If you meant to say that buildcache is an similar project with MSVC support, then yes: buildcache is similar but with different tradeoffs, feature sets and maturity. (And there is not only one alternative but several, for instance sccache which is much older than buildcache.) If you meant what you said literally, then no: buildcache is not a sibling ccache for Windows project. A sibling ccache for Windows project would be using the name "ccache" with the same configuration settings, etc., so that another project can use the same caching tool for several toolchains.  On the other hand, the "sibling ccache for Windows" idea has already been abandoned due to lack of interest (see #447). Adding MSVC support to the main ccache project is still possible, but that's not something I have bandwidth to drive personally.

Also: This PR/issue is about the implementation. Please comment in #431 (where you took my quoted comment from, not sure why you moved it here?) if you want to discuss whether the feature should be added at all.

rvogg commented:
What exactly is still missing for this feature to be integrated?
We are already using a fork of this branch productively, so we would be happy if it gets integrated into the master.
If it's blocked by things like the CI tests or something like that, we might be able to fix that.

cristianadam commented:
I've updated the MR on top of the 4.4 release.

I've added `/Zi` as too hard, this means that if you are using `Debug` and `RelWithDebInfo` will get cache misses, and get a correct build. 

In order to use ccache with `Debug` and `RelWithDebInfo` you need to replace `/Zi` with `/Z7`, which means that the debug information is stored in `.obj` files, and not in the compile pdb file, which is other compiler caching programs (sccache, buildcache) require.

To use is with CMake projects you need to have in your `CMakeLists.txt` somewhere:

```cmake
if (MSVC)
  foreach(config DEBUG RELWITHDEBINFO)
    foreach(lang C CXX)
      set(flags_var "CMAKE_${lang}_FLAGS_${config}")
      string(REPLACE "/Zi" "/Z7" ${flags_var} "${${flags_var}}")
    endforeach()
  endforeach()
endif()
```

Now it works with CMake and Ninja generator in `Release`, `Debug / RelWithDebInfo` with `/Z7`. I updated the top message with what's missing.

Next week I'll be away from keyboard, but when I come back I'll have a look at the missing items.

psi29a commented:
We over at OpenMW are waiting eagerly for this to land so we can collapse our 16 different MSVC builds down to 4. :)

jonesmz commented:
> We over at OpenMW are waiting eagerly for this to land so we can collapse our 16 different MSVC builds down to 4. :)

I'm curious what allows ccache to help with this. Could you elaborate?

psi29a commented:
> > We over at OpenMW are waiting eagerly for this to land so we can collapse our 16 different MSVC builds down to 4. :)
> 
> I'm curious what allows ccache to help with this. Could you elaborate?

OpenMW is a large FOSS project, which takes awhile to build. Bring down the time to compile is critical is important because our use-case is currently Gitlab shared runners that have build timeout set to about an hour. On anything other than windows, this is fine as we cache the results of ccache to be used so that we only compile files that are changed per push in MR or when merged, to our trunk.

On windows, specifically using MSVC, the timeout is 1 hour with a 15m bootup time to get all our deps in place for building leaving us with 45 minutes to fully build on every push. To fit within this budget, we've had to split up our builds to two different components (openmw, openmw-cs and tools), per build type (Release and Debug), per build system (msbuild and Ninja).

If we can finally get ccache for MSVC working, then we can behave like we do on all our other runners: android, debian, macos. Then we would only have to build files that have changed:
* cutting down the number of windows pipelines (builds) by half
* cutting down on build time
* being better 'tenants' for Gitlab as they host us freely (by being FOSS) by saving resources

rvjr commented:
> If we can finally get ccache for MSVC working, ...
Hi psi29a, we've also been using ccache a lot on Linux and wanted to use it on Windows as well. Since a few month we've been very happy with buildcache as a replacement. It worked for us with msbuild and ninja so far. The only modification we had to make was to replace pdb debug infos (/Zi) by built-in debug infos (/Z7).
In the end we would also love to use ccache on both platforms, just to have consistent tools, but this was a quick intermediate solution which might also help you.

jonesmz commented:
> On windows, specifically using MSVC, the timeout is 1 hour with a 15m bootup time to get all our deps in place for building leaving us with 45 minutes to fully build on every push. To fit within this budget, we've had to split up our builds to two different components (openmw, openmw-cs and tools), per build type (Release and Debug), per build system (msbuild and Ninja).

Thanks! This is what wasn't clear.

Merging several different builds of different parts of your source code into a single build makes sense as a ccache use case.

psi29a commented:
> Merging several different builds of different parts of your source code into a single build makes sense as a ccache use case.

Exactly, the sad thing about splitting up the work is that they all rely on our common 'component's library which is built for both openmw and openmw-cs builds, that's an extra 15m of duplicated work (having to recompile this every push). The next step would have been doing a staged pipeline to de-duplicate this work adding an overall delay of 20 minutes to the finished artefact and additional CI work. 

So having ccache there would simplify things quite a bit. :)

jrosdahl commented on src/ccache.cpp:
```diff
@@ -237,6 +238,9 @@ guess_compiler(string_view path)
   } else if (name == "pump" || name == "distcc-pump") {
     return CompilerType::pump;
+  } else if (name.find("cl") != nonstd::string_view::npos
+             || name.find("clang-cl") != nonstd::string_view::npos) {
```
Looking for "clang-cl" is redundant since if "clang-cl" is found in the string then "cl" has already been found in the string.

Perhaps we should check for an exact name instead of doing substring search?

cristianadam replied:
Ack.

jrosdahl commented:
@rvogg wrote:

> What exactly is still missing for this feature to be integrated?

There is no list of what is exactly missing, but some thoughts:

1. There is a need for somebody who wants to drive the implementation forward and answer my questions.
2. There is a need for a ccache maintainer with enough spare time, energy and motivation to review the changes. Currently I'm the only maintainer, and I reserve the right to take long mental breaks from non-bug-fix ccache issues. :slightly_smiling_face:
3. One thing that would help is if somebody else than me also would review and think about the proposed changes since I'm not very familiar with how stuff works in Windows land. And more generally, if people would help out with *other* ccache stuff (e.g. bug triaging, bug fixing, answering questions) then I would have more time to review and implement features.

jrosdahl commented:
@cristianadam: Thanks for the update. Regarding the unchecked checkboxes in the PR description:

> NMake Makefiles (requires usage of response files)

Any thoughts on the question _"Does this mean to write a parser similar to `args_init_from_gcc_atfile` but for an MSVC-specific format?"_ (from https://github.com/ccache/ccache/pull/506#issuecomment-584345009)?

What would happen if we merge the current PR without NMake Makefiles support? Will usage of NMake Makefiles with ccache just not work, or is there a risk that things may appear to work but misbehave silently (e.g. if you get false positive cache hits)?

> Precompile headers support (detection and storage of pch files)

Any thoughts on the question _"Should `/Yc` be too hard?"_ (from https://github.com/ccache/ccache/pull/506#issuecomment-584345009)? @rvjr thought so in https://github.com/ccache/ccache/pull/506#issuecomment-605336899 and also mentioned `/ZI`.

And similar here: What happens if we merge the current PR as is? Just a missing feature or is there a risk of things getting silently bad?

jrosdahl reviewed (changes requested):
Since the direct mode is not yet supported, I guess the code should switch off the direct mode if CompilerType::cl is detected?

rvogg commented on src/ccache.cpp:
```diff
@@ -931,9 +939,64 @@ to_cache(Context& ctx,
+  // MSVC compiler always print the input file name to stdout,
+  // plus parts of the warnings/error messages.
+  // So we have to fusion that into stderr...
+  // Transform \r\n into \n. This way ninja won't produce empty newlines
+  // for the /showIncludes argument.
```
> Could you expand on this a bit? What happens if `\r\n` are kept?

I think ccache from msvc build is generally broken here. 
I got part of the system tests working again on a private branch when ccache is built with mingw.
But when it is built with msvc, already the first tests fail because in the cached output (preprocessor, stdout, ...) there are always two line breaks instead of one.

cristianadam replied:
`/showInlcudes` is used by ninja to do its own dependency tracking. Without the conversion to `\n` ninja would insert newlines in the compilation output and would look like a mess.

With the new code to replace `\r\n` with `\n` also a MSVC build of ccache works as expected. The previous code relied on iostreams which are a bit broken on MSVC.

I tested with Qt Creator:
1. build with no cache, save the output of ninja
2. build with ccache, save the ouptut of ninja

Compare the two, warnings and other messages that the compiler produced were the same.

cristianadam replied:
Actually I used `> build.log 2>&1`, which is not the same as having stdout and stderr separate.

llunak replied:
The extra newlines come because STDERR_FILENO on Windows is open in text mode, which transforms newlines. Handled without a hack, I mean with a less of a hack, in #954 .

rvogg replied:
Commits b8c81b760935cf84c69b520ee266cbea22ce8ec3 and 60a6d2d51bcd6fb6e618c1be7aba933ae08d0a06 from #954  work very well. 
I tested them on top of #943 and was able to remove the hack in the run script.

I would suggest to cherry-pick them before integrating.

llunak replied:
I'm intentionally keeping #954 only as fixes for this, so no need to cherry-pick anything, it can go in right after this one at any time.

cristianadam commented on src/compopt.cpp:
```diff
@@ -140,6 +140,18 @@ const CompOpt compopts[] = {
+  {"/L", TAKES_ARG},                                                // msvc
+  {"/P", TOO_HARD},                                                 // msvc
+  {"/U", AFFECTS_CPP | TAKES_ARG | TAKES_CONCAT_ARG},               // msvc
+  {"/Zi", TOO_HARD},                                                // msvc
```
Yes. They also need to be added.

cristianadam commented on src/argprocessing.cpp:
```diff
@@ -510,7 +518,8 @@ process_arg(Context& ctx,
   // with gcc -E, when the output file is not specified.
-  if (args[i] == "-MD" || args[i] == "-MMD") {
+  if ((args[i] == "-MD" || args[i] == "-MMD")
+      && config.compiler_type() != CompilerType::cl) {
```
`-MD` needs to be checked, since it means ` /MD link with MSVCRT.LIB`. I'm not sure about `-MMD`.

cristianadam commented:
> @cristianadam: Thanks for the update. Regarding the unchecked checkboxes in the PR description:
> 
> > NMake Makefiles (requires usage of response files)
> 
> Any thoughts on the question _"Does this mean to write a parser similar to `args_init_from_gcc_atfile` but for an MSVC-specific format?"_ (from [#506 (comment)](https://github.com/ccache/ccache/pull/506#issuecomment-584345009))?
> 
> What would happen if we merge the current PR without NMake Makefiles support? Will usage of NMake Makefiles with ccache just not work, or is there a risk that things may appear to work but misbehave silently (e.g. if you get false positive cache hits)?

Some projects still use nmake or jom. I'm interested in ninja support, which works with CMake. I haven't tried msbuild.

> 
> > Precompile headers support (detection and storage of pch files)
> 
> Any thoughts on the question _"Should `/Yc` be too hard?"_ (from [#506 (comment)](https://github.com/ccache/ccache/pull/506#issuecomment-584345009))? @rvjr thought so in [#506 (comment)](https://github.com/ccache/ccache/pull/506#issuecomment-605336899) and also mentioned `/ZI`.
> 
> And similar here: What happens if we merge the current PR as is? Just a missing feature or is there a risk of things getting silently bad?

I'm fine with merging it like it is. We could document the pitfalls.

cristianadam commented on src/ccache.cpp:
```diff
@@ -931,9 +939,64 @@ to_cache(Context& ctx,
+      throw Failure(Statistic::bad_output_file);
+    }
+
+    Util::unlink_tmp(tmp_stderr2);
```
Done.

cristianadam commented:
Now that `NMake Makefiles` and `PCH` builds work, I'm happy about this MR!

llunak reviewed (commented):
>     1. There is a need for somebody who wants to drive the implementation forward and answer my questions.

Seeing that it's actually not that hard to get MSVC support for ccache, I'd like to see this working and start using it for LibreOffice builds. So I suppose I could do this to some extent.
 
>     2. There is a need for a ccache maintainer with enough spare time, energy and motivation to review the changes. Currently I'm the only maintainer, and I reserve the right to take long mental breaks from non-bug-fix ccache issues. slightly_smiling_face

I don't feel like being an actual maintainer for this. But I probably can do enough to keep this working for LibreOffice, similarly to what I've been doing for GCC/Clang PCH support.

>     3. One thing that would help is if somebody else than me also would review and think about the proposed changes since I'm not very familiar with how stuff works in Windows land.

I'm not exactly a Windows expert, but if this gets to the state when it becomes useful for building LibreOffice, then LO could start using ccache the same it's already the cache with GCC/Clang. I expect that'd be a reasonably good testcase.

> > Precompile headers support (detection and storage of pch files)

#952 adds support for using PCH files. As for storage:

> Any thoughts on the question _"Should `/Yc` be too hard?"

At least for now yes, since it wouldn't work as it is. It's even questionable if it should work, given how huge the PCH files can be, but since MSVC apparently cannot create reproducible PCH files, it actually might be useful (but then, CCACHE_PCH_EXTSUM could take care of that too and better). So the way I see it, if somebody wants it, they can implement it, but it's not that important.

> And similar here: What happens if we merge the current PR as is? Just a missing feature or is there a risk of things getting silently bad?

It seems to work at least in some way, and somebody mentioned above already using it in production. I haven't noticed anything particularly broken in the patch (barring the comments and the stdout/stderr mixing), so I think the worst that can happen is if somebody uses an option that ccache doesn't currently handle, which is really a similar situation to GCC/Clang support. I've also noticed error messages about failing stats file locks in the log file, which should need checking. Other than that, the basic stuff seems to work.

Would it be ok to accept the MSVC support at least as experimental? I understand that it can be annoying for you as the maintainer, and the support is nowhere near GCC/Clang support (e.g. no tests), but then adding MSVC support has apparently dragged on for years by now, and it's rather difficult to do more about it when it's in the form of pending pull requests. If not in master, then at least having this in a branch would IMO make it easier to get this finished. And you can always shrug problems off with "experimental, fix it yourself if you care enough".

rvogg commented on src/argprocessing.cpp:
```diff
@@ -262,7 +262,9 @@ process_arg(Context& ctx,
       ++argpath;
     }
-    auto file_args = Args::from_gcc_atfile(argpath);
+    auto file_args = ctx.config.compiler_type() == CompilerType::cl
```
As far as I know, the responsfile feature in gcc was taken over one to one from msvc (I can't find the link to the mailing list entry at the moment).
At this point you should simply rename the function to e.g. 'from_responsfile' or 'from_atfile' and call the same function in both cases. 
Particularly under Windows paths with blanks can occur, so you need the logic with the quoted arguments here.

rvogg commented on src/Args.cpp:
```diff
@@ -124,6 +124,19 @@ Args::from_gcc_atfile(const std::string& filename)
+    return nullopt;
+  }
+
+  return from_string(argtext);
```
'from_string' has no logic for arguments with spaces, it splits hard at whitespace charaters.

rvogg commented on src/argprocessing.cpp:
```diff
@@ -841,7 +853,8 @@ process_arg(Context& ctx,
-  if (args[i][0] == '-') {
+  if (args[i][0] == '-'
+      || (config.compiler_type() == CompilerType::cl && args[i][0] == '/')) {
     size_t slash_pos = args[i].find('/');
```
Under Windows it doesn't work at all, because it could be ill formed like '-includeC:/path/to/header.h'.

llunak replied:
On the other hand, this is not correct even on POSIX, as e.g. '-include../inc/header.h' won't work either.

Ccache's handling of compiler options has always been best-effort, and there always will be things that won't be handled correctly just because nobody's been bothered to make ccache handle this or that case properly. It's just a question of where the limit is set. And given the situation, I think in this case it'd make sense to lower the bar, and fix it up more in follow-up commits. Otherwise there's a risk this ends up like #162 all over again.

rvogg replied:
Several things are broken at this point. I have a more general solution to the problem on a branch, but I still need to prepare it for integration.
If @jrosdahl  is okay with that, I would also fix that in a later commit.

llunak commented on src/ccache.cpp:
```diff
@@ -931,9 +939,64 @@ to_cache(Context& ctx,
     return nonstd::make_unexpected(Statistic::missing_cache_file);
   }
 
+  // MSVC compiler always print the input file name to stdout,
+  // plus parts of the warnings/error messages.
+  // So we have to fusion that into stderr...
```
Done in #954 .

cristianadam commented:
@jrosdahl how should I proceed wit this MR? @llunak  has done some fixed in #952 and #954 which fix things that I haven't covered here.

The unresolved conversations have fixes in other MRs. Should I cherry-pick those fixes into this MR and mark the resolved conversations?

llunak commented:
I personally see no blocking issue with this PR. With my fixes what's pending is minor. I see only the two things pointed out by rvogg, and then the concatenated arguments with absolute paths problem, which is not even really a fault of this PR. All those can get fixed afterwards. I suggest accepting this PR and having the rest handled by new PRs from whatever cares to fix those.

> Since the direct mode is not yet supported, I guess the code should switch off the direct mode if CompilerType::cl is detected?

Is this something old? Direct mode seems to work fine. Or was this meant to say depend mode? That one gets disabled properly when this patch is used.

jrosdahl commented:
Some general thoughts on MSVC support:

While the changes in this PR are quite small, they will make the main ccache code have even more accidental complexity.

What I'm debating with myself at the moment is if I should accept even more complexity without a compiler abstraction, especially if the support comes without a test suite. Notably, there are wishes for supporting other compilers as well, e.g. #841/#942, and the ccache code will not cope well with just adding more and more complexity without improving code structure. I just created #956 for this.

A related problem is that it's hard for me to maintain support for compilers that either don't work on Linux or are not freely available. I would ideally want to be able to have tests for all supported compilers when developing locally, without having to rely too much on CI and especially not on having commercial compilers installed. That is, I would like a test suite that does not depend on external compilers. #957 created for this idea.

At the same time I don't want to stand in the way for people who are happy with the current state. This is actually very similar to what happened in #58 where a memcached backend was implemented. It worked reasonably well, but I couldn't stomach merging it because it would just be too unmaintainable and bolted on. It didn't end well; that work ended up on a branch which was removed eventually. (Then I found time and motivation to add proper storage backend support five years later.)

I would personally like to work on a proper compiler abstraction, but that would definitely take some time to finish, so the question is what to do with MSVC support in the meantime. Wait until compiler abstraction support is in place? Merge to branch? Merge to master (maybe under a feature flag) and try to live with the increased messiness?

I was leaning towards not merging this for now, but since @llunak has volunteered to help out with keeping it alive, I'm feeling more confident that I could live with this for a while.

In any case, I won't merge this before 4.5 since I would like to get some other changes out before thinking more about this.

jrosdahl commented:
@rvogg wrote:

> I think ccache from msvc build is generally broken here.

Note that building ccache itself with MSVC is experimental work-in-progress and nothing that is supported so far.

jrosdahl commented:
@llunak wrote:

> > There is a need for somebody who wants to drive the implementation forward and answer my questions.
>
> Seeing that it's actually not that hard to get MSVC support for ccache, I'd like to see this working and start using it for LibreOffice builds. So I suppose I could do this to some extent.

Thanks.

> > There is a need for a ccache maintainer with enough spare time, energy and motivation to review the changes. Currently I'm the only maintainer, and I reserve the right to take long mental breaks from non-bug-fix ccache issues. slightly_smiling_face
>
> I don't feel like being an actual maintainer for this. But I probably can do enough to keep this working for LibreOffice, similarly to what I've been doing for GCC/Clang PCH support.

Great!

> > Since the direct mode is not yet supported, I guess the code should switch off the direct mode if CompilerType::cl is detected?
>
> Is this something old? Direct mode seems to work fine. Or was this meant to say depend mode? That one gets disabled properly when this patch is used.

Don't remember, but it could be that I didn't think that MSVC's preprocessed output would be understood by ccache already.

jrosdahl commented:
> @jrosdahl how should I proceed wit this MR? @llunak has done some fixed in #952 and #954 which fix things that I haven't covered here.
> 
> The unresolved conversations have fixes in other MRs. Should I cherry-pick those fixes into this MR and mark the resolved conversations?

If I decide to merge MSVC support before #956, I think you can leave the current PR mostly as is and then we'll apply follow-up fixes separately.

rvogg commented:
@jrosdahl wrote:
> > I think ccache from msvc build is generally broken here.
>
> Note that building ccache itself with MSVC is experimental work-in-progress and nothing that is supported so far.

It was not meant to be harsh. With 'genarally' I wanted to say that with the msvc build the cached outputs always create two newlines, not only when cl is used. 
In #943  I had to add a hack to make the builds run green again. But as I said b8c81b7 and 60a6d2d fix the bug.

llunak commented:
> What I'm debating with myself at the moment is if I should accept even more complexity without a compiler abstraction, especially if the support comes without a test suite.

The important stuff (A level support code) comes with a test suite, so this PR breaking anything else is unlikely, and if it does then that's primarily responsibility of that existing test suite.  As for this PR not having tests, well, that's why it's B level, and it can change (and it's more likely to change if it's not made hard by hold this to the same standards as the existing A level code). Going from no MSVC support to MSVC support that maybe doesn't work in some cases is no regression. And in case you misssed it in #447 , I've built pretty much the entire LibreOffice including its bunch of 3rd-party libraries with PR plus my fixes on top.

> A related problem is that it's hard for me to maintain support for compilers that either don't work on Linux or are not freely available.

I thought the idea of B level was that you do not have to maintain support for it. Otherwise what's the point of the B level if there's no actual difference?

> I would ideally want to be able to have tests for all supported compilers when developing locally, without having to rely too much on CI and especially not on having commercial compilers installed.

MSVS has a community edition that is no less free than nvcc or icc. And you can test this on Linux too if you really want to, either clang-cl works anywhere where clang does (although it may need MS SDK headers, https://jake-shadle.github.io/xwin/) or apparently it's possible to use MSVC with Wine (https://github.com/mstorsjo/msvc-wine).

> I would personally like to work on a proper compiler abstraction, but that would definitely take some time to finish, so the question is what to do with MSVC support in the meantime. Wait until compiler abstraction support is in place? Merge to branch? Merge to master (maybe under a feature flag) and try to live with the increased messiness?

#956 may make some things cleaner, but it's not blocking for this. 'wc -l src/cache/*' is ~15kLOC and MSVC support is what, 300 out of that maybe? As for messiness, this PR has 8 compiler type checks and my latest source 13, and they are usually simple if blocks. For comparison, master has already ~20 compile_type checks. So it's not like this PR makes it way worse.
 
> In any case, I won't merge this before 4.5 since I would like to get some other changes out before thinking more about this.

You're the maintainer. I just hope you do realize that both asking people to take care of something and making it hard at the same time doesn't really work.

> @rvogg wrote:
> 
> > I think ccache from msvc build is generally broken here.
> 
> Note that building ccache itself with MSVC is experimental work-in-progress and nothing that is supported so far.

Works for me.

@cristianadam @rvogg : The only two actually unresolved issues in this PR are the @ file support (and even that is IMO hardly a blocker). Does either of you have (plans for) a follow-up commit to handle that?

cristianadam commented on src/compopt.cpp:
```diff
@@ -140,6 +140,20 @@ const CompOpt compopts[] = {
+  {"/U", AFFECTS_CPP | TAKES_ARG | TAKES_CONCAT_ARG},               // msvc
+  {"/Yc", TAKES_ARG | TOO_HARD},                                    // msvc
+  {"/ZI", TOO_HARD},                                                // msvc
+  {"/Zi", TOO_HARD},                                                // msvc
```
> @cristianadam @rvogg : The only two actually unresolved issues in this PR are the @ file support (and even that is IMO hardly a blocker). Does either of you have (plans for) a follow-up commit to handle that?

This MR support @file with the CMake's "NMakefile Generator", and in my test it compiled fine.

I haven't tested arguments with spaces, the only example I can think of would be a directory with spaces in it, source fiels with spaces are not very common.

I see this as a bug fix for later, than a requirement of having this MR accepted. I can have a look at this case, sure.

cristianadam commented:
> @cristianadam @rvogg : The only two actually unresolved issues in this PR are the @ file support (and even that is IMO hardly a blocker). Does either of you have (plans for) a follow-up commit to handle that?

This MR support @file with the CMake's "NMakefile Generator", and in my test it compiled fine.

I haven't tested arguments with spaces, the only example I can think of would be a directory with spaces in it, source fiels with spaces are not very common.

I see this as a bug fix for later, than a requirement of having this MR accepted. I can have a look at this case, sure.

rvogg commented:
> > @cristianadam @rvogg : The only two actually unresolved issues in this PR are the @ file support (and even that is IMO hardly a blocker). Does either of you have (plans for) a follow-up commit to handle that?
> 
> This MR support @file with the CMake's "NMakefile Generator", and in my test it compiled fine.
> 
> I haven't tested arguments with spaces, the only example I can think of would be a directory with spaces in it, source fiels with spaces are not very common.
> 
> I see this as a bug fix for later, than a requirement of having this MR accepted. I can have a look at this case, sure.

That is ok for me, I've already prepared the changes: https://github.com/cristianadam/ccache/pull/3, but I can also make a new pull request here.

llunak reviewed (approved):
As far as I can tell all the pointed out problems here have been fixed in #954 and cristianadam#3 , so from the technical side of things this PR should be ready.

jrosdahl commented:
@llunak wrote:

> 'wc -l src/cache/*' is ~15kLOC and MSVC support is what, 300 out of that maybe? As for messiness, this PR has 8 compiler type checks and my latest source 13, and they are usually simple if blocks. For comparison, master has already ~20 compile_type checks.

Please don't try to trivialize my concerns like this. That is very demotivating for me.

> So it's not like this PR makes it way worse.

It's exactly like this PR makes it worse, but only a bit. Yes, just adding a little more messiness just adds a little more messiness, but you can't keep doing that over and over again even if each addition just making things slightly messier.

jrosdahl commented:
I'll merge this after 4.5 is released, which I hope to be able to do in a couple of days.

llunak commented:
> Please don't try to trivialize my concerns like this. That is very demotivating for me.

Sorry. I didn't know 4.5 was meant to be out soon, so your comments here made it look to me like you were planning to postpone this PR for a long time after some rewrites that could possibly lead to all this needing to be rewritten all over again.

> I'll merge this after 4.5 is released, which I hope to be able to do in a couple of days.

I think it might make sense to keep a list of things needed for MSVC support, such as adding tests for this, so that we can sort all that out and not step on each others toes. I don't know where it'd be a good place to keep it, separate issues, wiki page or something else?

jrosdahl commented:
Thanks.

cristianadam commented:
> Thanks.

Thank you!

PJBoy commented:
Congrats on the merge. Will this be upgrading MSVC to a B on [the supported compilers](https://ccache.dev/platform-compiler-language-support.html) page?

jrosdahl commented:
> Congrats on the merge. Will this be upgrading MSVC to a B on [the supported compilers](https://ccache.dev/platform-compiler-language-support.html) page?

Yes (as already indicated on said page).

silverqx commented:
I would very much like to see this `Precompile headers support (detection and storage of pch files)` it's the only thing we are missing on msvc to use the ccache in everyday development. I'm already using the ccache in Github actions without any problem thx @cristianadam

silverqx commented:
qmake is using something like this for precompiled headers:
```
cl -c -FI..\include\pch.h -Yu..\include\pch.h -Fpdebug\TinyOrm_pch.pch ...
```

And cmake this:
```
cl.exe /YcE:/<path>/build-TinyOrm-Desktop_Qt_6_3_1_MSVC2022_64bit-Debug/CMakeFiles/TinyOrm.dir/cmake_pch.hxx 
/FpE:/<path>/build-TinyOrm-Desktop_Qt_6_3_1_MSVC2022_64bit-Debug/CMakeFiles/TinyOrm.dir/./cmake_pch.cxx.pch 
/FIE:/<path>/build-TinyOrm-Desktop_Qt_6_3_1_MSVC2022_64bit-Debug/CMakeFiles/TinyOrm.dir/cmake_pch.hxx
```

Trass3r commented:
The thing is that msvc's pch model is different from gcc's (better actually). They use a dummy source file and also produce an .obj file with code and debug info in addition to the pch file.

silverqx commented:
This is not good news if it is as you write, if msvc creates .obj for PCH in a different format than gcc so it means a new or customized caching logic will have to be created which doesn't exist now, which means more work. Build cache for dev. is really missing on msvc I have searched and didn't find anything working correctly.

Trass3r commented:
gcc doesn't produce code, only a .pch. Clang can be made to generate an .obj cause they need that for `clang-cl` mode, I think by now it's `-fpch-codegen -fpch-debuginfo`. Though it's also a bit special since you need to "compile" the .pch file after generating it, if my notes are still correct:
```sh
clang++ -x c++-header header.h -o header.pch -fpch-codegen -fpch-debuginfo
clang++ -c header.pch -o shared.o
```

akrieger commented:
Thank you for this. This cuts the github action time for our Windows CI more than in half, from a typical runtime of ~1:15 to about 25 minutes (15 minutes fixed cost to execute tests, 1 hour of setup + build time down to 10 minutes). That even includes folding a separate job into the same job to reuse build artifacts, so freeing up a whole runner slot per PR.
